scalar_udf.go

```diff
@@ -0,0 +1,247 @@
+
+void scalar_udf_callback(duckdb_function_info, duckdb_data_chunk, duckdb_vector);
+void scalar_udf_delete_callback(void *);
+
+typedef void (*scalar_udf_callback_t)(duckdb_function_info, duckdb_data_chunk, duckdb_vector);
+*/
+import "C"
+
+import (
+	"database/sql"
+	"database/sql/driver"
+	"runtime/cgo"
+	"unsafe"
+)
+
+type rowFn func(args []driver.Value) (any, error)
+
+type ScalarFuncConfig struct {
+	InputTypeInfos []TypeInfo
+	ResultTypeInfo TypeInfo
+
+	VariadicTypeInfo    *TypeInfo
+	Volatile            bool
+	SpecialNullHandling bool
+}
+
+type ScalarFuncExecutor struct {
+	RowExecutor rowFn
+}
+
+type ScalarFunc interface {
+	Config() ScalarFuncConfig
+	Executor() ScalarFuncExecutor
+}
+
+func setFuncError(function_info C.duckdb_function_info, msg string) {
+	err := C.CString(msg)
+	C.duckdb_scalar_function_set_error(function_info, err)
+	C.duckdb_free(unsafe.Pointer(err))
+}
+
+//export scalar_udf_callback
+func scalar_udf_callback(function_info C.duckdb_function_info, input C.duckdb_data_chunk, output C.duckdb_vector) {
+	extraInfo := C.duckdb_scalar_function_get_extra_info(function_info)
+
+	// extraInfo is a void* pointer to our ScalarFunc.
+	h := *(*cgo.Handle)(unsafe.Pointer(extraInfo))
+	function := h.Value().(ScalarFunc)
+
+	// Initialize the input chunk.
+	var inputChunk DataChunk
+	if err := inputChunk.initFromDuckDataChunk(input, false); err != nil {
+		setFuncError(function_info, getError(errAPI, err).Error())
+		return
+	}
+
+	// Initialize the output chunk.
+	var outputChunk DataChunk
+	if err := outputChunk.initFromDuckVector(output, true); err != nil {
+		setFuncError(function_info, getError(errAPI, err).Error())
+		return
+	}
+
+	// Execute the user-defined scalar function for each row.
+	executor := function.Executor()
+	values := make([]driver.Value, len(inputChunk.columns))
+	columnCount := len(values)
+
+	var err error
+	for rowIdx := 0; rowIdx < inputChunk.GetSize(); rowIdx++ {
+		// Set the values for each row.
+		for colIdx := 0; colIdx < columnCount; colIdx++ {
+			if values[colIdx], err = inputChunk.GetValue(colIdx, rowIdx); err != nil {
+				setFuncError(function_info, getError(errAPI, err).Error())
+				return
+			}
+		}
+
+		// Execute the function and write the result to the output vector.
+		var val any
+		if val, err = executor.RowExecutor(values); err != nil {
+			break
+		}
+		if err = outputChunk.SetValue(0, rowIdx, val); err != nil {
+			break
+		}
+	}
+
+	if err != nil {
+		setFuncError(function_info, getError(errAPI, err).Error())
+	}
+}
+
+//export scalar_udf_delete_callback
+func scalar_udf_delete_callback(extraInfo unsafe.Pointer) {
+	h := (*cgo.Handle)(extraInfo)
+	h.Delete()
+}
+
+func registerInputParams(config ScalarFuncConfig, f C.duckdb_scalar_function) error {
+	// Set variadic input parameters.
+	if config.VariadicTypeInfo != nil {
+		t := (*config.VariadicTypeInfo).logicalType()
+		C.duckdb_scalar_function_set_varargs(f, t)
+		C.duckdb_destroy_logical_type(&t)
+		return nil
+	}
```

Comment:
Would it be possible to register a function with both variadic args and non-variadic args?
If so, I don't think we shouldn't block users from registering with both variadic and regular args